File: String/Exercises.java
public class Exercises {
	
	/**
	* String - Implement strStr()
	*/
	public static int strStr(String haystack, String needle) {
        if (needle.isEmpty() && haystack.isEmpty()) {
            return 0;
        }
        if (needle.isEmpty()) {
            return 0;
        }
        if (haystack.isEmpty()) {
            return -1;
        }
        if (needle.length() > haystack.length()) {
			return -1;
		}

		int found = -1;
		char[] input = haystack.toCharArray();
		char[] seq = needle.toCharArray();
		char first = seq[0];

		for (int i = 0; i < input.length; i++) {
			if (first == input[i]) {
				int l = 0;
				for (int j = 0, k = i; j < seq.length && k < input.length; j++, k++) {
					if (seq[j] != input[k]) {
						break;
					} else {
						l++;
					}
				}
				if (l == seq.length) {
					found = i;
					break;
				}
			}
		}
		return found;
    }

    /**
    * String - Add Binary
    */
    public static String addBinary(String a, String b) {
        int carry = 0;
        char[] first = a.toCharArray();
        char[] second = b.toCharArray();
        int n = first.length;
        int m = second.length;
        int i = n - 1;
        int j = m - 1;
        StringBuilder sb = new StringBuilder();
        int x = Character.digit(first[i--], 2);
        int y = Character.digit(second[j--], 2);

        int sum = x + y + carry; 
        carry = sum / 2;
        int result = sum % 2;
        
        sb.append(result);
        
        while(i >= 0 || j >= 0) {
        	if (i >= 0) {
        		x = Character.digit(first[i--], 2);
        	} else {
        		x = 0;
        	}

        	if (j >= 0) {
        		y = Character.digit(second[j--], 2);
        	} else {
        		y = 0;
        	}
            sum = x + y + carry;
            carry = sum / 2;
            result = sum % 2;
            sb.append(result);
        }

        if (carry == 1) {
        	sb.append(1);
        }

        return sb.reverse().toString();
        }

	/**
	*String - Longest Common Prefix
	*/
	public static String longestCommonPrefix(String[] strs) {
        
        if(strs.length == 0) return "";
        
        String prefix = strs[0];
        
        for(int i = 1; i < strs.length; i++) {
            while(strs[i].indexOf(prefix) != 0) {
                prefix = prefix.substring(0, prefix.length() - 1);
                if (prefix.isEmpty()) return "";
            }
        }
        return prefix;
    }
	public static void main(String[] args) {

		// // Test String - Implement strStr()
		// String haystack = "Hello";
		// String needle = "ll";
		// System.out.println(strStr(haystack, needle));

		// // Test String - Implement Add Binary
		// String a = "111";
		// String b = "1";
		// System.out.println(addBinary(a, b));

		// Test String - Longest Common Prefix
		String[] test = {"c", "acc", "ccc"};
		String[] test1 = {"flag", "flirt", "flip"};
		System.out.println(longestCommonPrefix(test));
		System.out.println(longestCommonPrefix(test1));


	}
}

File: String/Exercises.py
class Solution:

    def strStr(self, haystack, needle):

        """
        String - Implement strStr()
        :type haystack: str
        :type needle: str
        :rtype: int
        """

        # if needle empty return 0
        if not needle:
            return 0
        # if haystack empty return -1
        if not haystack:
            return -1
        # if needle len > haystack len return -1
        if len(needle) > len(haystack):
            return -1

        found = -1
        first = needle[0]

        for i in range(len(haystack)):
            if haystack[i] == first:
                k = i
                match_len = 0
                for j in range(len(needle)):
                    if needle[j] != haystack[k]:
                        break
                    else:
                        match_len += 1
                    k += 1

                if match_len == len(needle):
                    found = i
                    break

        return found

    def addBinary(self, a, b):
        """
        String - Add Binary
        :type a: str
        :type b: str
        :rtype: str
        """
        i = len(a) - 1
        j = len(b) - 1
        result = []
        carry = 0
        a_lsb = a[i]
        b_lsb = b[j]
        while i >= 0 or j >= 0:
            if a_lsb == '1' and b_lsb == '1':
                if carry == 1:
                    result.append('1')
                else:
                    result.append('0')
                carry = 1
            elif a_lsb == '0' and b_lsb == '0':
                if carry == 1:
                    result.append('1')
                else:
                    result.append('0')
                carry = 0
            else:
                if carry == 1:
                    result.append('0')
                    carry = 1
                else:
                    result.append('1')
                    carry = 0
            i -= 1
            j -= 1
            a_lsb = a[i] if i >= 0 else '0'
            b_lsb = b[j] if j >= 0 else '0'
        if carry == 1:
            result.append('1')
        result.reverse()
        return ''.join(result)

    def longestCommonPrefix(self, strs):
        """
        String - Longest Common Prefix
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ""

        pattern = strs[0]

        for i in range(1, len(strs)):
            while strs[i].find(pattern) != 0:
                pattern = pattern[:-1]
        return pattern

if __name__ == '__main__':
    
    # # Test String - Implement strStr()
    # print(Solution().strStr("hello", 'll'))

    # # Test String - Add Binary
    # print(Solution().addBinary("111", "1"))

    # Test String - Longest Common Prefix
    print(Solution().longestCommonPrefix(["flower", "flight", "flame"]))
    print(Solution().longestCommonPrefix(["flower", "plane", "car"]))
    print(Solution().longestCommonPrefix(["c", "acc", "ccc"]))

